src/box/sql.c

```diff
@@ -1581,3 +1574,114 @@ sql_constraint_drop(uint32_t space_id, const char *name)
 	region_truncate(region, used);
 	return rc;
 }
+
+/**
+ * Create new constraint in space.
+ *
+ * @param name Name of the constraint.
+ * @param space_id ID of the space.
+ * @param path JSON-path of the new constraint.
+ * @param value Encoded value of the new constraint.
+ */
+static int
+sql_constraint_create(const char *name, uint32_t space_id, const char *path,
+		      const char *value)
+{
+	struct region *region = &fiber()->gc;
+	uint32_t used = region_used(region);
+	char key[16];
+	char *key_end = key + mp_format(key, 16, "[%u]", space_id);
+	/*
+	 * Even if there were no constraints of this type, it is possible that
+	 * _space contains a non-empty field of this type with an empty map as
+	 * its value, which affects the update operation.
+	 */
+	struct tuple *tuple;
+	if (box_index_get(BOX_SPACE_ID, 0, key, key_end, &tuple) != 0) {
+		region_truncate(region, used);
+		return -1;
+	}
+	assert(tuple != NULL);
+	uint32_t path_len = strlen(path);
+	uint32_t path_hash = field_name_hash(path, path_len);
+	const char *field = tuple_field_raw_by_full_path(
+		tuple_format(tuple), tuple_data(tuple), tuple_field_map(tuple),
+		path, path_len, path_hash, TUPLE_INDEX_BASE);
+	bool is_empty = field == NULL;
+
+	const char *ops;
+	size_t ops_size;
+	if (is_empty) {
+		ops = mp_format_on_region(region, &ops_size, "[[%s%s{%s%p}]]",
+					  "!", path, name, value);
+	} else {
+		uint32_t size = strlen(path) + strlen(name) + 2;
+		char *buf = region_alloc(region, size);
+		if (buf == NULL) {
+			region_truncate(region, used);
+			diag_set(OutOfMemory, size, "region_alloc", "buf");
+			return -1;
+		}
+		int len = snprintf(buf, size, "%s.%s", path, name);
+		assert(len >= 0 && (uint32_t)len < size);
+		(void)len;
+		ops = mp_format_on_region(region, &ops_size, "[[%s%s%p]]", "!",
+					  buf, value);
+	}
+	const char *end = ops + ops_size;
+	int rc = box_update(BOX_SPACE_ID, 0, key, key_end, ops, end, 0, NULL);
+	region_truncate(region, used);
+	return rc;
+}
+
+int
+sql_foreign_key_create(const char *name, uint32_t child_id, uint32_t parent_id,
+		       uint32_t child_fieldno, uint32_t parent_fieldno,
+		       const char *mapping)
+{
+	const struct space *child = space_by_id(child_id);
+	if (child == NULL) {
+		diag_set(ClientError, ER_NO_SUCH_SPACE, space_name);
+		return -1;
+	}
+	struct tuple_constraint_def *cdefs;
+	uint32_t count;
+	const char *path = NULL;
+	const char *value = NULL;
+
+	struct region *region = &fiber()->gc;
+	uint32_t used = region_used(region);
+	if (mapping == NULL) {
+		count = child->def->fields[child_fieldno].constraint_count;
+		cdefs = child->def->fields[child_fieldno].constraint_def;
+		path = tt_sprintf("format[%u].foreign_key", child_fieldno + 1);
```

Comment:
That is really dangerous. The static buffer better be used for very short living strings, but here it is supposed to survive `mp_format_on_region()` and the entire `sql_constraint_create()`. I would recommend to either store it on the region or on stack - this buffer's size is limited, 64 bytes would be more than enough given the max length of `%u`.